### DISCUSSION TITLE
How to wait first of multiple queue ?

### DISCUSSION THREAD

**Discussion by PaulBissemuth** [cocotb/cocotb#4009] [Q&A]
Hello all !

I'm new in the cocotb world, but exited about it. 
I try to test a module by taking advantage of the _test_matrix_multiplier_ example. 

Populate queues based on watchers like the _DataValidMonitor_ example seems pretty powerfull to me.
But I'm stuck in my model because I've **multiple queues** (based on a hdl generic) and **need to wait for the first one that is filled**. 

Example show a _await_ instruction for one queue, then another. I like this solution based on triggers to avoid an queue inspection each clock cycle (which should have devastating impact on simulation time). But it's sequential, how can I "_await_" concurrently ?
```
actual = await self.output_mon.values.get()
expected_inputs = await self.input_mon.values.get()
```

Usage of _cocotb.triggers.First()_ seems what I need but it expect triggers not coroutines.
```
first = cocotb.triggers.First( self.output_1.values.get(), self.output_2.values.get() )
TypeError: All triggers must be instances of Trigger! Got: coroutine"
```

Any help is welcome ! 

Thanks in advance

**Comment by ktbarrett** [comment]
This is kind of a sore point. The way `First` is written demands us to not support coroutines directly.

This is what you *can* do with `First`.
```python
async def select(*queues: Queue) -> Tuple[int, Any]:
    tasks = [cocotb.start_soon(q.get()) for q in queues]
    joins = [task.join() for task in tasks]
    first = await First(*joins)
    i = joins.index(first)
    # First leaves all tasks running, so we must kill the rest to prevent popping values
    for task in tasks:
        task.kill()
    return i, tasks[i].result()
```

I'm not *100%* sure this will work, but give it a go. You  might run into issues with things getting popped before you want, the Queue class really does need a public `wait_not_empty` method which would solve that.

This will be replaced with something that's easier to use in 2.0.
